File: back/src/main/java/com/back/standard/recommend/byRating/SimilarityRecommendByRating.java
package com.back.standard.recommend.byRating;

import com.back.standard.recommend.util.CosineSimilarityCalcer;
import com.back.standard.recommend.util.SimilarityCalcer;
import com.back.standard.recommend.util.Vector;

import java.util.*;
import static com.back.standard.recommend.util.Vector.VectorElement;
import static com.back.standard.recommend.util.SimilarityCalcer.Similar;

public class SimilarityRecommendByRating {

    public record Rating(
            long reviewerId,
            long subjectId,
            float rating
    ) { }

    private final Map<Long, Vector> ratingMatrix;
    private SimilarityCalcer calcer = new CosineSimilarityCalcer();

    public SimilarityRecommendByRating() {
        ratingMatrix = new HashMap<>();
    }

    public SimilarityRecommendByRating(SimilarityCalcer calcer) {
        ratingMatrix = new HashMap<>();
        this.calcer = calcer;
    }

    public void clear() {
        ratingMatrix.clear();
    }

    public void setData(List<Rating> reviews) {

        reviews.forEach(review -> {
            if (!ratingMatrix.containsKey(review.reviewerId())) {
                ratingMatrix.put(review.reviewerId(), new Vector());
            }
            ratingMatrix.get(review.reviewerId()).putValue(
                    review.subjectId(), review.rating());

        });

    }

    public List<Long> getRecommendList(long targetUserId, int referenceCnt, int maxRecommend) {

        Vector targetUser = ratingMatrix.getOrDefault(targetUserId, null);

        if (targetUser == null) return List.of();
        if (targetUser.isEmpty()) return List.of();

        ratingMatrix.replaceAll((_, oldValue) ->
                oldValue.subtractionValue(oldValue.getAverageValue()));

        List<Similar> similarList = calcer.getSimilarList(targetUser, ratingMatrix);

        Map<Long, Double> recommends = new HashMap<>();
        Set<Long> alreadyRead = targetUser.getLabels();

        for (int i = 0; i < referenceCnt && i < similarList.size(); i++) {

            double similarScore = similarList.get(i).score();

            List<VectorElement> compareReviewList =
                    ratingMatrix.get(similarList.get(i).id()).getVectorElementList();

            compareReviewList.stream()
                    .filter(rating -> !alreadyRead.contains(rating.label()))
                    .forEach(rating -> {
                        double score = rating.value() * similarScore;
                        if (recommends.containsKey(rating.label())
                                && recommends.get(rating.label()) <= score) {
                            return;
                        }
                        recommends.put(rating.label(), score);
                    });

        }

        return recommends
                .entrySet()
                .stream()
                .sorted((a, b) -> b.getValue().compareTo(a.getValue()))
                .limit(maxRecommend)
                .map(Map.Entry::getKey)
                .toList();
    }

}

File: back/src/main/kotlin/com/back/standard/recommend/byRating/SimilarityRecommendByRating.kt
package com.back.standard.recommend.byRating

import com.back.standard.recommend.util.CosineSimilarityCalcer
import com.back.standard.recommend.util.SimilarityCalcer
import com.back.standard.recommend.util.Vector

class SimilarityRecommendByRating<R, S>(
    private val calcer: SimilarityCalcer<R, S> = CosineSimilarityCalcer()
) {

    data class Rating<R, S>(
        val reviewerId: R,
        val subjectId: S,
        val rating: Float
    )

    private val ratingMatrix = mutableMapOf<R, Vector<S>>()

    fun clear() {
        ratingMatrix.clear()
    }

    fun setData(reviews: List<Rating<R, S>>) {
        reviews.forEach { review ->
            ratingMatrix.getOrPut(review.reviewerId) { Vector() }
                .putValue(review.subjectId, review.rating.toDouble())
        }
    }

    fun getRecommendList(targetUserId: R, referenceCnt: Int, maxRecommend: Int): List<S> {
        val targetUser = ratingMatrix[targetUserId] ?: return emptyList()
        if (targetUser.isEmpty()) return emptyList()

        ratingMatrix.values.forEach { it.subtractionValue(it.getAverageValue()) }

        val similarList = calcer.getSimilarList(targetUser, ratingMatrix)

        val recommends = mutableMapOf<S, Double>()
        val alreadyRead = targetUser.getLabels()

        for (i in 0 until minOf(referenceCnt, similarList.size)) {
            val similarScore = similarList[i].score

            val compareReviewList = ratingMatrix.getValue(similarList[i].id).getVectorElementList()

            compareReviewList
                .filter { it.label !in alreadyRead }
                .forEach { rating ->
                    val score = rating.value * similarScore
                    val existing = recommends[rating.label]
                    if (existing != null && existing <= score) return@forEach
                    recommends[rating.label] = score
                }
        }

        return recommends.entries
            .sortedByDescending { it.value }
            .take(maxRecommend)
            .map { it.key }
    }
}
